src/Cardano/Wallet.hs

```diff
@@ -263,28 +267,75 @@ mkWalletLayer db network = WalletLayer
-        -- there's only one thread per wallet that will apply blocks. And
-        -- therefore, only one thread making changes on checkpoints and/or tx
-        -- history
-        unsafeRunExceptT $ DB.putCheckpoint db (PrimaryKey wid) cp'
-        unsafeRunExceptT $ DB.putTxHistory db (PrimaryKey wid) txs
+    -- | Infinite restoration loop. We drain the whole available chain and try
+    -- to catch up with the node. In case of error, we log it and wait a bit
+    -- before retrying.
+    --
+    -- The function only terminates if the wallet has disappeared from the DB.
+    restoreStep :: WalletId -> (SlotId, SlotId) -> IO ()
+    restoreStep wid (slot, tip) = do
+        runExceptT (nextBlocks network slot) >>= \case
+            Left e -> do
+                TIO.putStrLn $ "[ERROR] restoreStep: " +|| e ||+ ""
+                restoreSleep wid slot
+            Right [] -> do
+                restoreSleep wid slot
+            Right blocks -> do
+                let next = slotId . header . last $ blocks
+                runExceptT (restoreBlocks wid blocks tip) >>= \case
+                    Left (ErrNoSuchWallet _) -> do
+                        TIO.putStrLn $ "[ERROR] restoreStep: wallet " +| wid |+ "is gone!"
+                    Right () -> do
+                        restoreStep wid (next, tip)
+
+    -- | Wait a short delay before querying for blocks again. We do take this
+    -- opportunity to also refresh the chain tip as it has probably increased
+    -- in order to refine our syncing status.
+    restoreSleep :: WalletId -> SlotId -> IO ()
+    restoreSleep wid slot = do
+        let tenSeconds = 10000000 in threadDelay tenSeconds
+        runExceptT (networkTip network) >>= \case
+            Left e -> do
+                TIO.putStrLn $ "[ERROR] restoreSleep: " +|| e ||+ ""
+                restoreSleep wid slot
+            Right (_, tip) ->
+                restoreStep wid (slot, slotId tip)
+
+    -- | Apply the given blocks to the wallet and update the wallet state,
+    -- transaction history and corresponding metadata.
+    restoreBlocks
+        :: WalletId
+        -> [Block]
+        -> SlotId -- ^ Network tip
+        -> ExceptT ErrNoSuchWallet IO ()
+    restoreBlocks wid blocks tip = do
+        let (inf, sup) =
+                ( slotId . header . head $ blocks
+                , slotId . header . last $ blocks
+                )
+        liftIO $ TIO.putStrLn $
+            "[INFO] Applying blocks ["+| inf |+" ... "+| sup |+"]"
+
+        (cp, meta) <- _readWallet wid
+        -- NOTE
+        -- We only process non-empty blocks, though we still keep the last block
+        -- of the list, even if empty, so that we correctly update the current
+        -- tip of the wallet state.
+        let nonEmpty = not . null . transactions
+        let (h,q) = first (filter nonEmpty) $ splitAt (length blocks - 1) blocks
+        let (txs, cp') = applyBlocks (h ++ q) cp
```

Comment:
isn't `(h,q)` first splitting `blocks` into two parts, then by `first` applying filter to both parts, and then when used as `(h++q)` the same as `filter nonEmpty blocks` ???

Reply:
ahh, no. `Control.Arrow` 's `first` will filter for the first group only... so q has one block and can be with or without transactions...but when it comes to applyBlocks if it has transactions belonging to us it has chances to be processed, but if not having transactions then no chances for sure...so why not use instead of `(h++q)` still `filter nonEmpty blocks`?

Reply:
As the note above indicates, we want to process the last block of the batch, always even if empty or irrelevant to us to update our current tip accordingly!